website/addons/owncloud/model.py

```diff
@@ -0,0 +1,153 @@
+from website.addons.base import exceptions
+from website.addons.base import StorageAddonBase
+from website.addons.base import (
+    AddonOAuthNodeSettingsBase, AddonOAuthUserSettingsBase,
+)
+from website.addons.owncloud.serializer import OwnCloudSerializer
+from website.addons.owncloud.settings import DEFAULT_HOSTS, USE_SSL
+from owncloud import Client as OwnCloudClient
+from website.addons.owncloud import settings
+
+from website.oauth.models import BasicAuthProviderMixin
+from website.util import api_v2_url
+
+logger = logging.getLogger(__name__)
+
+class OwnCloudProvider(BasicAuthProviderMixin):
+    """An alternative to `ExternalProvider` not tied to OAuth"""
+
+    name = 'ownCloud'
+    short_name = 'owncloud'
+    serializer = OwnCloudSerializer
+
+    def __repr__(self):
+        return '<{name}: {status}>'.format(
+            name=self.__class__.__name__,
+            status=self.account.display_name if self.account else 'anonymous'
+        )
+
+class AddonOwnCloudUserSettings(AddonOAuthUserSettingsBase):
+    oauth_provider = OwnCloudProvider
+    serializer = OwnCloudSerializer
+
+    def to_json(self, user):
+        ret = super(AddonOwnCloudUserSettings, self).to_json(user)
+        ret['hosts'] = DEFAULT_HOSTS
+        return ret
+
+class AddonOwnCloudNodeSettings(StorageAddonBase, AddonOAuthNodeSettingsBase):
+    oauth_provider = OwnCloudProvider
+    serializer = OwnCloudSerializer
+
+    folder_id = fields.StringField()
+
+    _api = None
+
+    @property
+    def api(self):
+        """authenticated ExternalProvider instance"""
+        if self._api is None:
+            self._api = OwnCloudProvider(self.external_account)
+        return self._api
+
+    @property
+    def folder_path(self):
+        return self.folder_id
+
+    @property
+    def folder_name(self):
+        return self.fetch_folder_name()
+
+    def set_folder(self, folder, auth=None):
+        if folder == '/ (Full ownCloud)':
+            folder = '/'
+        self.folder_id = folder
+        self.save()
+        if auth:
+            self.nodelogger.log(action='folder_selected', save=True)
+
+    def fetch_folder_name(self):
+        if self.folder_id == '/':
+            return '/ (Full ownCloud)'
+        return self.folder_id
+
+    def clear_settings(self):
+        self.folder_id = None
+
+    def deauthorize(self, auth=None, add_log=True):
+        """Remove user authorization from this node and log the event."""
+        self.clear_settings()
+        self.clear_auth()
+        if add_log and auth:
+            self.nodelogger.log(action='node_deauthorized', save=True)
+
+    def serialize_waterbutler_credentials(self):
+        if not self.has_auth:
+            raise exceptions.AddonError('Addon is not authorized')
+        provider = OwnCloudProvider(self.external_account)
+        return {'host': provider.host,
+                'username': provider.username,
+                'password': provider.password
+                }
```

Comment:
Cool. I'm applying this whenever I run into it, but if I miss a couple, let me know and I'll do another pass.